lib/checkassert.cpp

```diff
@@ -38,76 +38,59 @@ static const struct CWE CWE398(398U);   // Indicator of Poor Code Quality
 // Register this check class (by creating a static instance of it)
 namespace {
     CheckAssert instance;
-}
 
-void CheckAssert::assertWithSideEffects()
-{
-    if (!mSettings->severity.isEnabled(Severity::warning))
-        return;
+    struct ArgumentCheck {
+        const std::vector<const Token *> *arguments;
+        const Scope *assertionScope;
+    };
 
-    for (const Token* tok = mTokenizer->list.front(); tok; tok = tok->next()) {
-        if (!Token::simpleMatch(tok, "assert ("))
-            continue;
-
-        const Token *endTok = tok->next()->link();
-        for (const Token* tmp = tok->next(); tmp != endTok; tmp = tmp->next()) {
-            if (Token::simpleMatch(tmp, "sizeof ("))
-                tmp = tmp->linkAt(1);
+}
 
-            if (isVariableValueChangingOperator(tmp)) {
-                const auto *var = tmp->astOperand1()->variable();
-                if (checkVariableAssignmentSideEffect(var, tok->scope())) {
-                    assignmentInAssertError(tmp, var->name());
-                }
-            }
 
-            if (tmp->tokType() != Token::eFunction)   // TODO: constructor calls with initializer list are not detected!
-                continue;
-            const Function *func = tmp->function();
-            const auto args = getArguments(tmp);
-            if (isFunctionWithSideEffect(func, {&args, tok->scope()})) {
-                sideEffectInAssertError(tmp, func->name());
-            }
-        }
-        tok = endTok;
+/**
+ * \brief returns the lhs variable of the modification
+ * \param modifyOperator  - the modify operator
+ * \retval Variable - variable if found
+ * \retval nullptr  - if no variable found
+ */
+static const Variable *getLhsVariable(const Token *modifyOperator) {
+    const Token *lhs = modifyOperator->astOperand1();
+    while (lhs != modifyOperator && lhs->variable() == nullptr) {
+        // de-referenced or address of -> this seems to be something with a side effect
+        // let checkArgument decide
+        if (lhs->str() == "*" || lhs->str() == "&")
+            lhs = lhs->next();
```

Comment:
`lhs` is conditionally changed in the loop.. that is dangerous because the loop will hang if `lhs` is not changed for some reason.

I do not think it's reasonable to use next here. Imagine: `x[0] = 0;` if the `modifyOperator->astOperand1()` is the `[` there will be no variable returned? and if we have `x[y] = z` then this function will claim that `y` is assigned?